leader/directed_ring/peterson.go

```diff
@@ -0,0 +1,173 @@
+package directed_ring
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/krzysztof-turowski/distributed-framework/lib"
+	"log"
+)
+
+type statePeterson struct {
+	Max    int
+	Status modeType
+}
+
+type messagePeterson struct {
+	Max  int
+	Mode modeType
+}
+
+func initializePeterson(v lib.Node) bool {
+	s := statePeterson{Max: v.GetIndex(), Status: unknown}
+	data, _ := json.Marshal(s)
+	v.SetState(data)
+	return false
+}
+
+func sendMessagePeterson(v lib.Node, m messagePeterson) {
+	outMessage, _ := json.Marshal(m)
+	v.SendMessage(0, outMessage)
+}
+
+func receiveMessagePeterson(v lib.Node) messagePeterson {
+	var m messagePeterson
+	inMessage := v.ReceiveMessage(0)
+	json.Unmarshal(inMessage, &m)
+	return m
+}
+
+func getStatePeterson(v lib.Node) statePeterson {
+	var s statePeterson
+	json.Unmarshal(v.GetState(), &s)
+	return s
+}
+
+func setStatePeterson(v lib.Node, s statePeterson) {
+	data, _ := json.Marshal(s)
+	v.SetState(data)
+}
+
+func processPeterson(v lib.Node, round int) bool {
+	s := getStatePeterson(v)
+	if s.Status != relay {
+		if s.Status == unknown {
+			sendMessagePeterson(v, messagePeterson{Max: s.Max, Mode: unknown})
+			m1 := receiveMessagePeterson(v)
+			if m1.Mode == leader {
+				s.Max = m1.Max
+				s.Status = nonleader
+				sendMessagePeterson(v, messagePeterson{Max: m1.Max, Mode: leader})
+			} else if m1.Max == v.GetIndex() {
+				s.Status = leader
+				sendMessagePeterson(v, messagePeterson{Max: s.Max, Mode: leader})
+			} else {
+				if s.Max > m1.Max {
+					sendMessagePeterson(v, messagePeterson{Max: s.Max, Mode: unknown})
+				} else {
+					sendMessagePeterson(v, messagePeterson{Max: m1.Max, Mode: unknown})
+				}
+				m2 := receiveMessagePeterson(v)
+				if m2.Mode == leader {
+					s.Max = m2.Max
+					s.Status = nonleader
+					sendMessagePeterson(v, messagePeterson{Max: m2.Max, Mode: leader})
+				} else if m2.Max == v.GetIndex() {
+					s.Status = leader
+					sendMessagePeterson(v, messagePeterson{Max: s.Max, Mode: leader})
+				} else if m1.Max >= s.Max && m1.Max >= m2.Max {
+					s.Max = m1.Max
+				} else {
+					s.Status = relay
+				}
+			}
+			setStatePeterson(v, s)
+			if s.Status == nonleader {
+				return true
+			}
```

Comment:
Please remove this and just add "return true " to the respective ifs (line 59 and 72 if I see correctly)

Reply:
leads to adding setStatePeterson(v, s) in lines 59 and 72 too